File: algoritmo_genetico.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <random>
#include <numeric>

using namespace std;

// Estrutura para armazenar as coordenadas de uma cidade
struct Cidade {
    double x, y;
};

// Hiperparâmetros
const int TAM_POPULACAO = 100;
const int MAX_GERACOES = 1000;
const double TAXA_MUTACAO = 0.05;
const double TAXA_CROSSOVER = 0.90;
const int TAM_TORNEIO = 3;

// Função para calcular a distância Euclidiana entre duas cidades
double calcular_distancia(Cidade a, Cidade b) {
    return sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2));
}

// Calcula a distância total de uma rota
double calcular_distancia_total(const vector<int>& rota, const vector<Cidade>& cidades) {
    double totalDist = 0;
    for (size_t i = 0; i < rota.size() - 1; ++i) {
        totalDist += calcular_distancia(cidades[rota[i]], cidades[rota[i+1]]);
    }
    // Retorna à cidade de origem
    totalDist += calcular_distancia(cidades[rota.back()], cidades[rota.front()]);
    return totalDist;
}

// Inicializa a população aleatoriamente
vector<vector<int>> inicializar_populacao(int numcidades, mt19937& rng) {
    vector<vector<int>> populacao(TAM_POPULACAO, vector<int>(numcidades));
    vector<int> rota_base(numcidades);
    iota(rota_base.begin(), rota_base.end(), 0); 

    for (int i = 0; i < TAM_POPULACAO; ++i) {
        populacao[i] = rota_base;
        shuffle(populacao[i].begin(), populacao[i].end(), rng);
    }
    return populacao;
}

// Seleção por Torneio
vector<int> selecao_torneio(const vector<vector<int>>& populacao, const vector<double>& fitnesses, mt19937& rng) {
    uniform_int_distribution<int> dist(0, TAM_POPULACAO - 1);
    int melhor_Idx = dist(rng);
    
    for (int i = 1; i < TAM_TORNEIO; ++i) {
        int idx = dist(rng);
        if (fitnesses[idx] < fitnesses[melhor_Idx]) { // Menor distância = melhor fitness
            melhor_Idx = idx;
        }
    }
    return populacao[melhor_Idx];
}

// Crossover Ordenado (OX)
vector<int> ordenar_crossover(const vector<int>& pai1, const vector<int>& pai2, mt19937& rng) {
    int numcidades = pai1.size();
    uniform_int_distribution<int> dist(0, numcidades - 1);
    
    int inicio = dist(rng);
    int fim = dist(rng);
    if (inicio > fim) swap(inicio, fim);

    vector<int> filho(numcidades, -1);
    vector<bool> em_filho(numcidades, false);

    // Copia um trecho do Pai 1
    for (int i = inicio; i <= fim; ++i) {
        filho[i] = pai1[i];
        em_filho[pai1[i]] = true;
    }

    // Preenche o resto com o Pai 2
    int index_atual = (fim + 1) % numcidades;
    for (int i = 0; i < numcidades; ++i) {
        int p2Cidade = pai2[(fim + 1 + i) % numcidades];
        if (!em_filho[p2Cidade]) {
            filho[index_atual] = p2Cidade;
            index_atual = (index_atual + 1) % numcidades;
        }
    }
    return filho;
}

// Mutação por Troca (Swap)
void mutar(vector<int>& rota, mt19937& rng) {
    uniform_real_distribution<double> probDist(0.0, 1.0);
    if (probDist(rng) < TAXA_MUTACAO) {
        uniform_int_distribution<int> idxDist(0, rota.size() - 1);
        int idx1 = idxDist(rng);
        int idx2 = idxDist(rng);
        swap(rota[idx1], rota[idx2]);
    }
}

int main() {
    int numcidades;
    if (!(cin >> numcidades)) return 0; // Lê a primeira linha: número de cidades

    vector<Cidade> cidades(numcidades);
    for (int i = 0; i < numcidades; ++i) {
        cin >> cidades[i].x >> cidades[i].y; // Lê as coordenadas x e y
    }

    random_device rd;
    mt19937 rng(rd());

    vector<vector<int>> populacao = inicializar_populacao(numcidades, rng);
    
    double melhor_distancia_global = numeric_limits<double>::max();
    vector<int> melhor_rota_global;

    for (int generation = 0; generation < MAX_GERACOES; ++generation) {
        vector<double> distancias(TAM_POPULACAO);
        
        // Avaliação
        for (int i = 0; i < TAM_POPULACAO; ++i) {
            distancias[i] = calcular_distancia_total(populacao[i], cidades);
            if (distancias[i] < melhor_distancia_global) {
                melhor_distancia_global = distancias[i];
                melhor_rota_global = populacao[i];
            }
        }

        vector<vector<int>> nova_populacao;
        
        // Elitismo: mantém o melhor da geração
        nova_populacao.push_back(melhor_rota_global);

        // Gera o restante da população
        uniform_real_distribution<double> probDist(0.0, 1.0);
        while (nova_populacao.size() < TAM_POPULACAO) {
            vector<int> pai1 = selecao_torneio(populacao, distancias, rng);
            vector<int> pai2 = selecao_torneio(populacao, distancias, rng);
            
            vector<int> filho;
            if (probDist(rng) < TAXA_CROSSOVER) {
                filho = ordenar_crossover(pai1, pai2, rng);
            } else {
                filho = pai1;
            }

            mutar(filho, rng);
            nova_populacao.push_back(filho);
        }

        populacao = nova_populacao;
    }

    cout << "Melhor Distancia Encontrada: " << melhor_distancia_global << endl;
    cout << "Rota: ";
    for (int CidadeIdx : melhor_rota_global) {
        cout << CidadeIdx << " -> ";
    }
    cout << melhor_rota_global.front() << " (Retorno)" << endl;

    return 0;
}

File: algoritmo_genetico.py
import math
import random


# =========================================================
# HIPERPARÂMETROS - CONFIGURAÇÃO FORTE
# =========================================================
TAM_POPULACAO = 80
MAX_GERACOES = 600
TAXA_MUTACAO = 0.08
TAXA_CROSSOVER = 0.85
TAM_TORNEIO = 4

# =========================================================
# HIPERPARÂMETROS - CONFIGURAÇÃO FRACA
# =========================================================
# TAM_POPULACAO = 40
# MAX_GERACOES = 200
# TAXA_MUTACAO = 0.02
# TAXA_CROSSOVER = 0.75
# TAM_TORNEIO = 2


# =========================================================
# CLASSE
# =========================================================

class AlgoritmoGenetico:

    def __init__(self, coordenadas):

        self.coordenadas = coordenadas
        self.numcidades = len(coordenadas)

        self.historico_melhor = []
        self.historico_medio = []

        self.melhor_distancia_global = float("inf")
        self.melhor_rota_global = None

    # =====================================================
    # DISTÂNCIA EUCLIDIANA
    # =====================================================

    def calcular_distancia(self, a, b):

        return math.sqrt(
            (a[0] - b[0])**2 +
            (a[1] - b[1])**2
        )

    # =====================================================
    # DISTÂNCIA TOTAL
    # =====================================================

    def calcular_distancia_total(self, rota):

        total = 0.0

        for i in range(len(rota) - 1):

            cidade_a = self.coordenadas[rota[i]]
            cidade_b = self.coordenadas[rota[i + 1]]

            total += self.calcular_distancia(cidade_a, cidade_b)

        cidade_final = self.coordenadas[rota[-1]]
        cidade_inicio = self.coordenadas[rota[0]]

        total += self.calcular_distancia(cidade_final, cidade_inicio)

        return total

    # =====================================================
    # POPULAÇÃO INICIAL
    # =====================================================

    def inicializar_populacao(self):

        populacao = []

        rota_base = list(range(self.numcidades))

        for _ in range(TAM_POPULACAO):

            nova_rota = rota_base[:]

            random.shuffle(nova_rota)

            populacao.append(nova_rota)

        return populacao

    # =====================================================
    # SELEÇÃO POR TORNEIO
    # =====================================================

    def selecao_torneio(self, populacao, distancias):

        melhor_idx = random.randint(0, TAM_POPULACAO - 1)

        for _ in range(1, TAM_TORNEIO):

            idx = random.randint(0, TAM_POPULACAO - 1)

            if distancias[idx] < distancias[melhor_idx]:

                melhor_idx = idx

        return populacao[melhor_idx]

    # =====================================================
    # CROSSOVER OX
    # =====================================================

    def ordenar_crossover(self, pai1, pai2):

        numcidades = len(pai1)

        inicio = random.randint(0, numcidades - 1)
        fim = random.randint(0, numcidades - 1)

        if inicio > fim:
            inicio, fim = fim, inicio

        filho = [-1] * numcidades

        em_filho = [False] * numcidades

        # Copia trecho do pai1
        for i in range(inicio, fim + 1):

            filho[i] = pai1[i]

            em_filho[pai1[i]] = True

        # Completa com pai2
        index_atual = (fim + 1) % numcidades

        for i in range(numcidades):

            cidade_p2 = pai2[(fim + 1 + i) % numcidades]

            if not em_filho[cidade_p2]:

                filho[index_atual] = cidade_p2

                index_atual = (index_atual + 1) % numcidades

        return filho

    # =====================================================
    # MUTAÇÃO
    # =====================================================

    def mutar(self, rota):

        if random.random() < TAXA_MUTACAO:

            idx1 = random.randint(0, len(rota) - 1)
            idx2 = random.randint(0, len(rota) - 1)

            rota[idx1], rota[idx2] = rota[idx2], rota[idx1]

    # =====================================================
    # EXECUÇÃO PRINCIPAL
    # =====================================================

    def executar(self, verbose=True):

        populacao = self.inicializar_populacao()

        for geracao in range(MAX_GERACOES):

            distancias = [
                self.calcular_distancia_total(rota)
                for rota in populacao
            ]

            # =============================================
            # MELHOR SOLUÇÃO
            # =============================================

            for i in range(TAM_POPULACAO):

                if distancias[i] < self.melhor_distancia_global:

                    self.melhor_distancia_global = distancias[i]

                    self.melhor_rota_global = populacao[i][:]

            # =============================================
            # HISTÓRICO
            # =============================================

            self.historico_melhor.append(
                self.melhor_distancia_global
            )

            self.historico_medio.append(
                sum(distancias) / len(distancias)
            )

            # =============================================
            # NOVA POPULAÇÃO
            # =============================================

            nova_populacao = []

            # Elitismo
            nova_populacao.append(
                self.melhor_rota_global[:]
            )

            while len(nova_populacao) < TAM_POPULACAO:

                pai1 = self.selecao_torneio(
                    populacao,
                    distancias
                )

                pai2 = self.selecao_torneio(
                    populacao,
                    distancias
                )

                # Crossover
                if random.random() < TAXA_CROSSOVER:

                    filho = self.ordenar_crossover(
                        pai1,
                        pai2
                    )

                else:

                    filho = pai1[:]

                # Mutação
                self.mutar(filho)

                nova_populacao.append(filho)

            populacao = nova_populacao

            if verbose:

                print(
                    f"Geração {geracao+1:4d} | "
                    f"Melhor: {self.melhor_distancia_global:.2f} | "
                    f"Média: {self.historico_medio[-1]:.2f}"
                )

        return (
            self.melhor_rota_global,
            self.melhor_distancia_global
        )